### ISSUE TITLE
calledWith with functions arguments not recognized

### ISSUE THREAD

**Issue by omirobarcelo** [timkindberg/jest-when#50]
I have this function  
```
  watchDataInformationRequests(
    sortFn?: (a: DataInformationRequest, b: DataInformationRequest) => number
  ): Observable<DataInformationRequest[]> {
    return this.store.select(RequestsState.dataInformationRequests(sortFn));
  }
```
and I want to mock the `store.select` when it receives a `RequestsState.dataInformationRequests(sortFn)`.

I tried:
```
const fromNewestToOldest = (a: { createdAt: Date }, b: { createdAt: Date }) =>
  b.createdAt.getTime() - a.createdAt.getTime();

when(store.select)
  .calledWith(RequestsState.dataInformationRequests(fromNewestToOldest))
  .mockReturnValue(of(sortedRequests));
when(store.select)
  .calledWith(RequestsState.dataInformationRequests(expect.anything()))
  .mockReturnValue(of(sortedRequests));
when(store.select)
   .calledWith(expect.any(RequestsState.dataInformationRequests))
  .mockReturnValue(of(sortedRequests));
when(store.select)
  .calledWith(RequestsState.dataInformationRequests)
  .mockReturnValue(of(sortedRequests));
```  
In all the cases, when I call `watchDataInformationRequests` the when is not recognized, and it returns null. 

I am trying to do an integration test, so I cannot just mock `watchDataInformationRequests`. 

I tried mocking `store.select` directly, and in that case it works, so I imagine that the problem comes from `when` not being able to properly compare that the functions are the same, or I am providing the wrong value to `when`.

How can I mock a function when it expects another function as argument?

**Comment by timkindberg** [comment]
Hmm well I'd think if the function is the same function reference it would work. Does `console.log(RequestsState.dataInformationRequests(fromNewestToOldest) === RequestsState.dataInformationRequests(fromNewestToOldest))` ouput `true`?

**Comment by omirobarcelo** [comment]
No, it outputs `false`.  
And `fromNewestToOldest` is only defined in the test, the component passes the same function, but probably with a different reference, that's why I tried:
```
when(store.select)
  .calledWith(RequestsState.dataInformationRequests(expect.anything()))
  .mockReturnValue(of(sortedRequests));
```

**Comment by timkindberg** [comment]
Hmm so I think that's the problem. What about a test like this:

```
const fn1 = RequestsState.dataInformationRequests(fromNewestToOldest)
const fn2 = RequestsState.dataInformationRequests(fromNewestToOldest)

expect(fn1).toBe(fn2)
expect(fn1).toEqual(fn2)
```

I'm expecting both of those `expects` to fail. If so, then we can't really expect jest-when to understand that they are "equal" either. So it's behaving inline with how normal jest works. I'm not really sure how we could test the equality of functions that are not referentially equivalent.

TLDR; it needs to be the same actual function reference for this to work.

**Comment by omirobarcelo** [comment]
Both assertions failed. Then I assume there is no way to do this, since it's not possible to decide that the arguments are the same?

**Comment by timkindberg** [comment]
Correct. Not possible. Think about it: have you ever done this before in your own code? Nice avatar btw.
